### DISCUSSION TITLE
Compatibility Issue: must save mesh before loading it with pyvista, inefficient workload

### DISCUSSION THREAD

**Discussion by Hendrik-code** [pyvista/pyvista#4532] [Q&A]
Hey
If I am not mistaken, pyvista lacks compatibility with nearly every other mesh-related package.

My current problem:
I want to use numpy-stl to create the mesh, then use it with pyvista. The caveat is: I have to save the mesh as stl and then load it for pyvista, because I currently don't see a way for pyvista to eat my mesh object created with numpy-stl.
(This is similiar to o3d as well)

Is there a good workaround for this?
I really don't need the meshes saved on the disk, and even though I could delete them after the program is done, this would still be ugly and inefficient.

**Comment by banesullivan** [comment]
PyVista is built on NumPy and, for many mesh data structures, has a straightforward interface for creating new meshes from NumPy arrays.

Can you provide some example code with `numpy-stl` or the other packages you use? If they are using numpy arrays, the transition from that library's object to a PyVista object should be trivial in most cases -- something we'd be happy to help you implement if you can share with us an example to get started.

On your note about o3d, this library has come up in a few examples in the past. Here are some examples where we translate an o3d object to a PyVista object.

- https://github.com/pyvista/pyvista-support/issues/170#issuecomment-639211683
- https://github.com/pyvista/pyvista-support/issues/178#issuecomment-643066518
- https://github.com/pyvista/pyvista/discussions/1589#discussioncomment-4060645
- https://github.com/pyvista/pyvista/discussions/1405

Additionally, many external libraries have streamlined interoperability with PyVista -- typically in the form of a `to_pyvista()` method implemented in the other library rather than have `from_<object_type>()` method in PyVista. One that may be relevant to you is [PyntCloud](https://github.com/daavoo/pyntcloud). 

Otherwise, PyVista has direct support for [meshio](https://github.com/nschloe/meshio) and you can use `meshio` objects as an intermediate, in-memory translation layer with PyVista's `from_meshio()` method as meshio implements a standard interface for defining mesh objects.

**Reply by MatthewFlamm** [reply]
I would temper the response a bit by saying that while _possible_ to construct new meshes from numpy arrays, it requires knowledge of vtk cell structure for `PolyData` and `UnstructuredGrids`. For complex cell types like Polyhedral, it quickly becomes something non-trivial, although definitely still workable if you have the knowledge to do so.  A lot of the examples you provided were for pointset-like data which I would agree is fairly straightforward to construct directly.

I think PyVista can be better about this in general.  It should work "out-of-the-box" when possible.  A limitation for PyVista is that only some vtk Reader classes will support in-memory loading.  Some of the vtk Readers do support it, but it isn't exposed in PyVista's Reader API.  For example, the vtk XML readers all support binary in-memory reading as implemented in #1778, but PyVista does not expose this possibility outside of pickling.  Wherever possible we should allow something like:

```py
reader = pv.XMLPolyDataReader(inmemory_object_representing_data)
reader.read()
```

But for STL, I do not think vtk supports this same functionality, so I would also recommend `meshio`.  If desired, a feature request for vtk could be opened to allow PyVista to use it in the future.
